File: templates/code-carl/ch09_dp/t22_lc0121_max_profit.py
"""LeetCode 121. Best Time to Buy and Sell Stock
https://leetcode.cn/problems/best-time-to-buy-and-sell-stock/
"""

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.maxProfit([7, 1, 5, 3, 6, 4]) == 5
    assert solution.maxProfit([7, 6, 4, 3, 1]) == 0
    print("All tests passed.")

File: solutions/java/code-carl/ch09_dp/T22_LC0121_MaxProfit.java
/**
 * Practice template for:
 * - LeetCode 121. Best Time to Buy and Sell Stock
 *   https://leetcode.cn/problems/best-time-to-buy-and-sell-stock/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/23 23:22
 * @comment
 */
public class T22_LC0121_MaxProfit {
    public int maxProfit(int[] prices) {
        int[] dp = new int[2];
        dp[0] = -prices[0];
        dp[1] = 0;
        for (int i = 1; i < prices.length; i++) {
            dp[0] = Math.max(dp[0], -prices[i]);
            dp[1] = Math.max(dp[1], dp[0] + prices[i]);
        }
        return dp[1];
    }
}
